forgerock-auth/src/main/java/org/forgerock/android/auth/callback/AppIntegrityCallback.kt

```diff
@@ -0,0 +1,199 @@
+
+private val TAG = AppIntegrityCallback::class.java.simpleName
+
+/**
+ * Callback to collect the device binding information
+ */
+open class AppIntegrityCallback : NodeAware, AbstractCallback {
+
+    @Keep
+    @JvmOverloads
+    constructor(jsonObject: JSONObject, index: Int) : super(jsonObject, index)
+
+    @Keep
+    @JvmOverloads
+    constructor() : super()
+
+    /* Currently not support Support Standard Integrity
+    companion object {
+        val cache = mutableMapOf<String, StandardIntegrityTokenProvider>()
+    }
+     */
+
+    private lateinit var node: Node
+
+    lateinit var requestType: RequestType
+        private set
+
+    /**
+     * The projectNumber received from server
+     */
+    lateinit var projectNumber: String
+        private set
+
+    /**
+     * The nonce received from server
+     */
+    lateinit var nonce: String
+        private set
+
+    final override fun setAttribute(name: String, value: Any) = when (name) {
+        "requestType" -> requestType = RequestType.valueOf((value as String).uppercase())
+        "projectNumber" -> projectNumber = value as String
+        "nonce" -> nonce = value as String
+        else -> {}
+    }
+
+    override fun getType(): String {
+        return "AppIntegrityCallback"
+    }
+
+    /**
+     * Input the Token to the server
+     * @param value The JWS value.
+     */
+    fun setToken(value: String) {
+        super.setValue(value, 0)
+    }
+
+    /**
+     * Input the Client Error to the server
+     * @param value DeviceBind ErrorType .
+     */
+    fun setClientError(value: String) {
+        super.setValue(value, 1)
+    }
+
+    /**
+     * Request for Integrity Token from Google SDK
+     *
+     * @param context  The Application Context
+     * @param listener The Listener to listen for the result
+     */
+    open fun requestIntegrityToken(context: Context,
+                                   listener: FRListener<Void>) {
+        val scope = CoroutineScope(Dispatchers.Default)
+        scope.launch {
+            try {
+                requestIntegrityToken(context)
+                Listener.onSuccess(listener, null)
+            } catch (e: Exception) {
+                Listener.onException(listener, e)
+            }
+        }
+    }
+
+    /**
+     * Bind the device.
+     *
+     * @param context  The Application Context
+     */
+    open suspend fun requestIntegrityToken(context: Context) {
+
+        try {
+            when (requestType) {
+                RequestType.CLASSIC -> {
+                    val integrityManager = getIntegrityManager(context)
+                    val builder = IntegrityTokenRequest.builder()
+                    builder.setCloudProjectNumber(projectNumber.toLong())
+                    builder.setNonce(hashWithNonce(getAuthId()))
+                    setToken(integrityManager.requestIntegrityToken(builder.build()).await()
+                        .token())
+                }
+
+                /*
```

Comment:
is this for future implementation ?

Reply:
Yes, this is new from Google, would like to get it ready. I have high-level tested it, and it works as expected, but not ready to make it official at this moment.